### PR TITLE
Use Token instead of login

### PR DESCRIPTION
**Pull request by KobbyMmo**
## Description

To shorten Vercel function request TTL, use WordPress token instead of WordPress username and password. We only have [10 seconds ](https://vercel.com/docs/platform/limits)to run any serverless function, so calls to get preview times out if it takes longer.
image

Fixes # (N/A)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by kilemensi — COMMENTED** [review]
👍🏾 @KobbyMmo 

QQ: Do ACF fields also get "revision-ed"?

**Review by nerdlet — APPROVED** [review]

File: src/lib/wp/index.js [hunk 1]
```diff
@@ -362,20 +357,27 @@ function wp(site) {
...
       revisionId,
       thumbnailId,
+      token,
       locale = siteServer.defaultLocale,
     }) => ({
       get page() {
         return (async () => {
           if (id && revisionId) {
-            return getPageRevisionById(id, revisionId, thumbnailId, locale);
+            return getPageRevisionById(id, revisionId, token, locale);
```
[the remark this answers is not in the document]
Comment (KobbyMmo) [reply src/lib/wp/index.js]: we do

File: src/pages/api/preview.js [hunk 2]
```diff
@@ -4,10 +4,11 @@ export default async function preview(req, res) {
     postId,
     revisionId,
     _thumbnail_id: thumbnailId,
+    token,
   } = req.query;
 
   res.setPreviewData({
-    query: { id: postId, revisionId, thumbnailId },
+    query: { id: postId, revisionId, thumbnailId, token },
```
Comment (KhadijaMahanga) [inline src/pages/api/preview.js:11]: We are not receiving `thumbnailId` as part of the preview query

**Review by KhadijaMahanga — COMMENTED** [review]
@KobbyMmo we are not receiving `thumbnailId` from the preview query. So my local breaks at `article.thumbnail_image`.

Did we miss adding it on dashboard?

We can also get the `featured_media` from the result json object (resource) like how we are retrieving `author`. But also double check if `featured_media` is included on post revision. Because I am getting empty on my local wordpress, though its on my post. https://developer.wordpress.org/rest-api/reference/post-revisions/

File: src/lib/wp/index.js [hunk 3]
```diff
@@ -246,11 +229,21 @@ function wp(site) {
...
+      revisionId,
+      token,
+      lang
+    );
     if (isEmpty(resource)) {
       return resource;
     }
+    const thumbnail = await getResourceById("media", thumbnailId, lang);
+
+    resource.featured_image_src = thumbnail?.source_url || null;
```
Comment (KhadijaMahanga) [inline src/lib/wp/index.js:245]: `thumbnailId` is undefined

File: src/lib/wp/index.js [hunk 4]
```diff
@@ -246,11 +229,21 @@ function wp(site) {
...
+    const resource = await getRevisionById(
+      "pages",
+      id,
+      revisionId,
+      token,
+      lang
+    );
     if (isEmpty(resource)) {
       return resource;
     }
```
Comment (KhadijaMahanga) [inline src/lib/wp/index.js:242]: If empty?

**Comment by KhadijaMahanga** [comment]
> 👍🏾 @KobbyMmo
> 
> QQ: Do ACF fields also get "revision-ed"?

@kilemensi included as part of post revision https://support.advancedcustomfields.com/forums/topic/revisions-of-custom-fields/

**Comment by KobbyMmo** [comment]
@KhadijaMahanga , I realized no featured image is on your post is not on your post, I added one, and it worked alright

File: src/pages/api/preview.js (hunk 2 above)
Comment (KobbyMmo) [reply src/pages/api/preview.js:11]: its attached by WP

**Comment by KhadijaMahanga** [comment]
@KobbyMmo all dashboard post preview are going to preview error.

**Comment by KhadijaMahanga** [comment]
> @KhadijaMahanga , I realized no featured image is on your post is not on your post, I added one, and it worked alright. The featured image is not attached to the post, instead, it is attached as a thumbnail Id, I will investigate manually attaching it in WP

Feature image is attached as thumbnail id on the WP preview query? This is what I am getting on my preview url `http://localhost:3000/api/preview/?postType=posts&postId=6&revisionId=7&token=<token>`  and the one on dashboard looks the same `https://promisetracker.dev.codeforafrica.org/api/preview/?postType=posts&postId=1612&revisionId=1613&token=<token>` 


Also all the post preview on dashboard are going to `preview error`, which one worked alrright?

**Comment by KobbyMmo** [comment]
>This is what I am getting on my preview url

@KhadijaMahanga  I have realized they are going to preview error, it's still exceeding the time limit, so I am trying to attach all the possible things we will need ie author, thumbnailSrc. When I test on my local its 7 seconds but it exceeds it on vercel.

> Also all the post preview on dashboard are going to `preview error`, which one worked alright?

 For the _thumbnail_id,  this is what I get for your TestArticle `https://promisetracker.dev.codeforafrica.org/api/preview/?postType=posts&postId=1612&revisionId=1613&token=<token>&preview=true&_thumbnail_id=1484` 

I am even handling null images on the frontend, so with or without images, it shouldn't throw an error.


If this doesn't work it might mean we will not be able to view previews on Vercel, maybe we will have to switch production to Dokku

**Pull request merged**